[PL-46512] Add custom delegate logging info (#5135)
harness/developer-hub: docs/platform/delegates/manage-delegates/create-custom-delegate-logs.md

        <then>
            <appender name="REST2" class="io.harness.logging.remote.RemoteStackdriverLogAppender">
                <threshold>TRACE</threshold>
                <managerHost>${MANAGER_HOST_AND_PORT}</managerHost>
                <accountId>${ACCOUNT_ID}</accountId>
                <clientCertPath>${DELEGATE_CLIENT_CERTIFICATE_PATH:- }</clientCertPath>
                <clientCertKey>${DELEGATE_CLIENT_CERTIFICATE_KEY_PATH:- }</clientCertKey>
                <trustAllCerts>${TRUST_ALL_CERTIFICATES:-false}</trustAllCerts>
                <logStreamingUrl>${LOG_STREAMING_SERVICE_URL:-${MANAGER_HOST_AND_PORT}/log-service/}</logStreamingUrl>
                <delegateToken>${DELEGATE_TOKEN:-${ACCOUNT_SECRET}}</delegateToken>
                <appName>delegate</appName>
            </appender>
        </then>
    </if>

    <logger name="software.wings" level="${LOGGING_LEVEL:-INFO}"/>
    <logger name="org.zeroturnaround" level="WARN"/>
    <logger name="io.harness.pcf" level="${LOGGING_LEVEL_PCF:-INFO}"/>
    <logger name="io.harness.event.client.impl" level="${LOGGING_LEVEL_EVENT_CLIENT:-INFO}"/>
    <logger name="io.github.resilience4j" level="WARN"/>

>     // todo(avmohan): Temporarily added till we root cause the high frequency logging issue

Reviewer: Do we want to leave these todos in the example?

Author: Nice catch @SushrutHarness. Removing. Thanks.

---
title: Create custom delegate logs
description: This topic describes how to create custom delegate logs.
sidebar_position: 8
---

> The delegate automatically creates a new log daily, named `delegate.log`. You can create custom delegate logs if you want to collect additional logging information.

Author: ```suggestion
The delegate automatically creates a new log daily, named `delegate.log`. You can customize delegate logging if the default setup doesn't fit your needs. For example, you can customize the layout, verbosity, and destination of the messages.
```

For more information on default delegate logs, go to [Delegate logs](/docs/platform/delegates/delegate-concepts/delegate-overview#delegate-logs).

> To create a custom delegate log for Kubernetes and Docker delegates, you can provide a custom `logback.xml` file to the delegate. You can accomplish this by mounting the file inside the delegate container or building it in your custom container. Then, update the delegate `JAVA_OPTS` with the logback option for the custom path to the configuration. This will enable you to customize the logging behavior of the delegate according to your specific needs.

Author: ```suggestion
To create customized delegate logging for Kubernetes and Docker delegates, you can provide a custom `logback.xml` file to the delegate. You can accomplish this by mounting the file inside the delegate container or building it in your custom container. Then, update the delegate `JAVA_OPTS` with the logback option for the custom path to the configuration. This will enable you to customize the logging behavior of the delegate according to your specific needs.
```

:::info
You can configure Kubernetes and Docker delegates in various ways. Below is an example Kubernetes delegate configuration.

:::

Let's look at an example Kubernetes delegate configuration.

To create a custom custom delegate log, do the following:

1. Create a ConfigMap containing your custom `logback.xml` file.

   ```
   # This creates a new ConfigMap named custom-logback in a harness-delegate-ng namespace
   kubectl create configmap custom-logback -n harness-delegate-ng --from-file=custom-logback.xml
   ```

2. Update your `delegate.yaml` file to mount the new ConfigMap.

   ```yaml
        volumeMounts:
        - name: config-volume
          mountPath: /opt/harness-delegate/logback/
      volumes:
      - name: config-volume
        configMap:
          name: custom-logback
          items:
          - key: custom-logback.xml
            path: custom-logback.xml
   ```

> 3. Update your `delegate.yaml` file with you new `JAVA_OPTS` value.

Author: ```suggestion
3. Update your `delegate.yaml` file with your new `JAVA_OPTS` value.
```

   ```yaml
        - name: JAVA_OPTS
          value: "-Dlogback.configurationFile=/opt/harness-delegate/logback/custom-logback.xml"

   ```

### Default logging configuration

Here is the default logging configuration for the Harness Delegate.

:::info

> The following configurations added by Harness to the default Logback xml.

Author: ```suggestion
The following configurations were added by Harness to the default Logback XML.
```

        <then>
            <appender name="REST2" class="io.harness.logging.remote.RemoteStackdriverLogAppender">
                <threshold>TRACE</threshold>
                <managerHost>${MANAGER_HOST_AND_PORT}</managerHost>
                <accountId>${ACCOUNT_ID}</accountId>
                <clientCertPath>${DELEGATE_CLIENT_CERTIFICATE_PATH:- }</clientCertPath>
                <clientCertKey>${DELEGATE_CLIENT_CERTIFICATE_KEY_PATH:- }</clientCertKey>
                <trustAllCerts>${TRUST_ALL_CERTIFICATES:-false}</trustAllCerts>
                <logStreamingUrl>${LOG_STREAMING_SERVICE_URL:-${MANAGER_HOST_AND_PORT}/log-service/}</logStreamingUrl>
                <delegateToken>${DELEGATE_TOKEN:-${ACCOUNT_SECRET}}</delegateToken>
                <appName>delegate</appName>
            </appender>
        </then>
    </if>

    <logger name="software.wings" level="${LOGGING_LEVEL:-INFO}"/>
    <logger name="org.zeroturnaround" level="WARN"/>
    <logger name="io.harness.pcf" level="${LOGGING_LEVEL_PCF:-INFO}"/>
    <logger name="io.harness.event.client.impl" level="${LOGGING_LEVEL_EVENT_CLIENT:-INFO}"/>
    <logger name="io.github.resilience4j" level="WARN"/>

> 

Author: ```suggestion
```

> 

Author: ```suggestion
```

        <then>
            <appender name="REST2" class="io.harness.logging.remote.RemoteStackdriverLogAppender">
                <threshold>TRACE</threshold>
                <managerHost>${MANAGER_HOST_AND_PORT}</managerHost>
                <accountId>${ACCOUNT_ID}</accountId>
                <clientCertPath>${DELEGATE_CLIENT_CERTIFICATE_PATH:- }</clientCertPath>
                <clientCertKey>${DELEGATE_CLIENT_CERTIFICATE_KEY_PATH:- }</clientCertKey>
                <trustAllCerts>${TRUST_ALL_CERTIFICATES:-false}</trustAllCerts>
                <logStreamingUrl>${LOG_STREAMING_SERVICE_URL:-${MANAGER_HOST_AND_PORT}/log-service/}</logStreamingUrl>
                <delegateToken>${DELEGATE_TOKEN:-${ACCOUNT_SECRET}}</delegateToken>
                <appName>delegate</appName>
            </appender>
        </then>
    </if>

    <logger name="software.wings" level="${LOGGING_LEVEL:-INFO}"/>
    <logger name="org.zeroturnaround" level="WARN"/>
    <logger name="io.harness.pcf" level="${LOGGING_LEVEL_PCF:-INFO}"/>
    <logger name="io.harness.event.client.impl" level="${LOGGING_LEVEL_EVENT_CLIENT:-INFO}"/>
    <logger name="io.github.resilience4j" level="WARN"/>

>     // todo(avmohan): Temporarily added till we root cause the high frequency logging issue

Author: ```suggestion
```

    <logger name="io.kubernetes.client.informer.cache.ReflectorRunnable" level="${KUBE_WATCH_LEVEL:-OFF}"/>

>     // todo: Temporarily added till we replace fabric8 dependency

Author: ```suggestion
```
